cli/lib/service-generator.ts

```diff
@@ -0,0 +1,344 @@
+const REG_API = /((get|post|put|delete)[a-zA-Z]+):\s*{\n\s*api:\s*'(\/api(\/:?[a-zA-Z-]+)+)'/g;
+
+const getBasicTypeData = (props: JSONSchema & { propertyName?: string }) => {
+  const { type, properties = {}, items, required = [] } = props || {};
+  let value;
+
+  if (type === 'object') {
+    const data: { [p: string]: string | object } = {};
+    // eslint-disable-next-line guard-for-in
+    for (const key in properties) {
+      const pData = properties[key] as JSONSchema;
+      const __key = required.includes(key) ? key : `${key}?`;
+      data[__key] = getBasicTypeData({ ...pData, propertyName: key });
+    }
+    value = data;
+  } else if (type === 'array' && items) {
+    const itemsType = get(items, 'type');
+    if (itemsType === 'object' || itemsType === 'array') {
+      value = `Array<${formatJson(getSchemaData(items))}>`;
+    } else {
+      value = `${itemsType}[]`;
+    }
+  } else if (type === 'integer' || type === 'number') {
+    value = 'number';
+  } else {
+    value = type;
+  }
+  return value;
+};
+
+const formatJson = (data: string | object) => {
+  const jsonData = typeof data === 'string' ? data : JSON.stringify(data, null, 2);
+  return jsonData.replace(/\\+n/g, '\n').replace(/"/g, '');
+};
+
+const getSchemaData: any = (schemaData: JSONSchema) => {
+  const { $ref, type } = schemaData || {};
+  let res;
+  if ($ref) {
+    const quoteList = $ref.split('/').slice(1);
+    res = getSchemaData(get(swaggerData, quoteList));
+  } else if (type) {
+    res = getBasicTypeData(schemaData);
+  } else {
+    res = schemaData;
+  }
+  return res || {};
+};
+
+const getResponseType = (schemaData: JSONSchema) => {
+  const { $ref } = schemaData || {};
+  if ($ref) {
+    const quoteList = $ref.split('/').slice(1);
+    const _data = get(swaggerData, [...quoteList, 'properties', 'data']) || {};
+    if (_data.type === 'object' && _data.properties?.total && _data.properties?.list) {
+      return 'pagingList';
+    } else {
+      return _data.type;
+    }
+  } else {
+    return schemaData?.type;
+  }
+};
+
+const formatApiPath = (apiPath: string) => {
+  const pathParams: { [p: string]: string } = {};
+  const newApiPath = apiPath.replace(/(:[a-zA-Z]+)/g, (p: string) => {
+    const pName = p.slice(1);
+    pathParams[pName] = pName.toLocaleLowerCase().indexOf('id') >= 0 ? 'number' : 'string';
+    return `<${pName}>`;
+  });
+  return {
+    url: newApiPath,
+    pathParams,
+  };
+};
+
+const writeSteadyContent = (content: string, filePath: string) => {
+  const lastIndex = content.indexOf('GENERATED');
+  let steadyContent = content;
+  steadyContent = lastIndex ? content.slice(0, lastIndex + 9) : content;
+
+  fs.writeFileSync(filePath, `${steadyContent}\r\n`, 'utf8');
+};
+
+const autoGenerateService = (
+  content: string,
+  filePath: string,
+  isEnd: boolean,
+  resolve: (value: void | PromiseLike<void>) => void,
+) => {
+  // only deal with suffix of '-api.ts'
+  if (!filePath.endsWith('-api.ts') && !isEnd) {
```

Comment:
what does `&& !isEnd` mean here?

Reply:
if the current file's suffix is not '-api.ts', and the file search is not finished, it will continue and deal with the next file

Reply:
It's useless here, this flag is used for do something at the whole process end. Just check the suffix is ok.